File: radix-sort/radix-sort.rs
fn radix_sort(arr: &mut Vec<u32>) {
    let mut max = *arr.iter().max().unwrap();
    let mut exp = 1;
    while max / exp > 0 {
        counting_sort(arr, exp);
        exp *= 10;
    }
}

fn counting_sort(arr: &mut Vec<u32>, exp: u32) {
    let mut output = vec![0; arr.len()];
    let mut count = vec![0; 10];

    for &num in arr.iter() {
        count[((num / exp) % 10) as usize] += 1;
    }

    for i in 1..10 {
        count[i] += count[i - 1];
    }

    for &num in arr.iter().rev() {
        let idx = ((num / exp) % 10) as usize;
        count[idx] -= 1;
        output[count[idx]] = num;
    }

    arr.copy_from_slice(&output);
}

fn main() {
    let mut arr = (0..20).map(|_| rand::random::<u32>() % 20).collect::<Vec<u32>>();
    let start_time = std::time::Instant::now();
    radix_sort(&mut arr);
    let duration = start_time.elapsed();
    println!("Vreme izvršenja: {} ms", duration.as_millis());
}

File: radix-sort/radix-sort.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

void counting_sort(int *arr, int n, int exp);
void radix_sort(int *arr, int n);

void counting_sort(int *arr, int n, int exp)
{
    int *output = (int *)malloc(n * sizeof(int));
    int count[10] = {0};

#pragma omp parallel for
    for (int i = 0; i < n; i++)
    {
#pragma omp atomic
        count[(arr[i] / exp) % 10]++;
    }

    for (int i = 1; i < 10; i++)
    {
        count[i] += count[i - 1];
    }

#pragma omp parallel for
    for (int i = n - 1; i >= 0; i--)
    {
        int idx = --count[(arr[i] / exp) % 10];
        output[idx] = arr[i];
    }

#pragma omp parallel for
    for (int i = 0; i < n; i++)
    {
        arr[i] = output[i];
    }

    free(output);
}

void radix_sort(int *arr, int n)
{
    int max = arr[0];

    for (int i = 1; i < n; i++)
    {
        if (arr[i] > max)
            max = arr[i];
    }

    for (int exp = 1; max / exp > 0; exp *= 10)
    {
        counting_sort(arr, n, exp);
    }
}

int main()
{
    int n = 20;
    int *arr = (int *)malloc(n * sizeof(int));

    for (int i = 0; i < n; i++)
    {
        arr[i] = rand() % n;
    }

    double start_time = omp_get_wtime();
    radix_sort(arr, n);
    double end_time = omp_get_wtime();

    printf("Vreme izvršenja: %f ms\n", (end_time - start_time) * 1000);

    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("\n");

    free(arr);
    return 0;
}